src/FileHelper.php

```diff
@@ -335,21 +342,100 @@ public static function copyDirectory(string $source, string $destination, array
 
             if ($filter === null || $filter->match($from)) {
                 if (is_file($from)) {
-                    if (!$destinationExists) {
-                        self::ensureDirectory($destination, $dirMode);
-                        $destinationExists = true;
-                    }
-                    copy($from, $to);
-                    if ($fileMode !== null) {
-                        chmod($to, $fileMode);
-                    }
+                    self::copyFile($from, $to, $options);
                 } elseif ($recursive) {
                     self::copyDirectory($from, $to, $options);
                 }
             }
         }
 
         closedir($handle);
+
+        self::processCallback($afterCopy, $source, $destination);
+    }
+
+    /**
+     * Copies files with some options.
+     *
+     * - dirMode: integer or null, the permission to be set for newly copied directories. Defaults to null.
+     *   When null - directory will be not created
+     * - fileMode: integer, the permission to be set for newly copied files. Defaults to the current environment
+     *   setting.
+     * - beforeCopy: callback, a PHP callback that is called before copying file. If the callback
+     *   returns false, the copy operation for file will be cancelled. The signature of the
+     *   callback should be: `function ($from, $to)`, where `$from` is the file to be copied from,
+     *   while `$to` is the copy target.
+     * - afterCopy: callback, a PHP callback that is called after file if successfully copied.
+     *   The signature of the callback should be: `function ($from, $to)`, where `$from` is the file
+     *   copied from, while `$to` is the copy target.
+     *
+     * @param string $source The source file
+     * @param string $destination The destination filename
+     * @param array $options
+     *
+     * @psalm-param array{
+     *   dirMode?: int,
+     *   fileMode?: int,
+     *   beforeCopy?: callable,
+     *   afterCopy?: callable,
+     * } $options
+     */
+    public static function copyFile(string $source, string $destination, array $options = []): void
+    {
+        if (!is_file($source)) {
+            throw new InvalidArgumentException('Argument $source must be an existing file.');
+        }
+
+        $dirname = dirname($destination);
+        $dirMode = $options['dirMode'] ?? null;
+        $fileMode = $options['fileMode'] ?? null;
+        $afterCopy = $options['afterCopy'] ?? null;
+        $beforeCopy = $options['beforeCopy'] ?? null;
+
+        if (self::processCallback($beforeCopy, $source, $destination) === false) {
+            return;
+        }
+
+        if (!is_dir($dirname)) {
+            if ($dirMode === null) {
+                throw new RuntimeException('Directory ' . $dirname . ' does not exist and cannot be created.');
+            }
```

Comment:
Why don't can create directory if not defined dir mode?

Reply:
> Why don't can create directory if not defined dir mode?

I was try repeat native `copy` function by default

Reply:
This is change previous behavior: if dir mode not set, then use 0755 (see [here](https://github.com/yiisoft/files/blob/0f0b76ebadd0493d3a30dd257ab6269248439116/src/FileHelper.php#L308)).